File: LeetCode/Python/CopyListWithRandomPointer.py
__author__ = 'yi'


class RandomListNode:
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None


class Solution:
    def copyRandomList(self, head):
        copy = head
        while copy is not None:
            nextnode = copy.next
            copy.next = RandomListNode(copy.label)
            copy.next.next = nextnode
            copy = nextnode

        copy = head
        while copy is not None:
            copy.next.random = copy.random.next
            copy = copy.next.next

        copy = head
        head_copy = copy.next if copy is not None else None

        while copy is not None:
            next_node = copy.next.next
            copy.next.next = next_node.next
            copy.next = next_node
            copy = next_node
        return head_copy

File: LeetCode/Java/CopyListWithRandomPointer.java
import java.util.HashMap;
import java.util.Map;

/**
 * Created by yi on 4/23/15.
 */
public class CopyListWithRandomPointer {
    public RandomListNode copyRandomList(RandomListNode head) {
        RandomListNode dummyHead = new RandomListNode(0);
        RandomListNode copy = head;
        RandomListNode temp = dummyHead;
        Map<RandomListNode, RandomListNode> ref = new HashMap<RandomListNode, RandomListNode>();
        while (copy != null) {
            temp.next = new RandomListNode(copy.label);
            ref.put(copy, temp.next);
            temp = temp.next;
            copy = copy.next;
        }
        copy = head;
        temp = dummyHead;
        while (copy != null) {
            temp.next.random = ref.get(copy.random);
            temp = temp.next;
            copy = copy.next;
        }
        return dummyHead.next;
    }

    public RandomListNode copyRandomList1(RandomListNode head) {
        RandomListNode ori = head;
        while (ori != null) {
            RandomListNode next = head.next;
            ori.next = new RandomListNode(ori.label);
            ori.next.next = next;
            ori = next;
        }
        ori = head;
        while (ori != null) {
            ori.next.random = ori.random != null ? ori.random.next : null;
            ori = ori.next.next;
        }
        ori = head;
        RandomListNode copyHead = ori.next != null ? ori.next : null;
        while (ori != null) {
            RandomListNode copy = ori.next;
            ori.next = copy.next;
            ori = ori.next;
            copy.next = ori != null ? ori.next : null;
        }
        return copyHead;

    }
}
